File: 백준/Silver/11725. 트리의 부모 찾기/트리의 부모 찾기.py
"""
[트리, DFS]
1. 노드 1부터 dfs를 활용
2. 탐색하면서 다음 노드의 부모 노드를 현재 노드로 설정
"""

import sys
input = sys.stdin.readline

N = int(input())
graph = dict()
for _ in range(N-1):
    u, v = map(int, input().split())
    if u not in graph:
        graph[u] = []
    if v not in graph:
        graph[v] = []
    graph[u].append(v)
    graph[v].append(u)

ans = [0]*(N+1)
stack = []
visited = [False]*(N+1)
stack.append(1)
visited[1] = True

while stack:
    u = stack.pop()
    for v in graph[u]:
        if not visited[v]:
            visited[v] = True
            stack.append(v)
            ans[v] += u
for i in range(2, N+1):
    print(ans[i])

File: 백준/Silver/11725. 트리의 부모 찾기/트리의 부모 찾기.java
import java.io.*;
import java.util.*;

/*
 * 1. 전부 인접 리스트로 연결.
 * 2. 1부터 dfs 돌면서 노드 배열에 부모 저장 후 출력
 */
public class Main {
	static int N, parent[];
	static List<Integer> list[];
	static boolean visited[];
    public static void main(String[] args) throws Exception {
    		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    		N = Integer.parseInt(br.readLine());
    		list = new ArrayList[N+1];
    		parent = new int[N+1];
    		for(int i=1; i<=N; i++)
    			list[i] = new ArrayList<>();
    		
    		for(int i=0; i<N-1; i++) {
    			StringTokenizer st = new StringTokenizer(br.readLine());
    			int node = Integer.parseInt(st.nextToken());
    			int next = Integer.parseInt(st.nextToken());
    			list[node].add(next);
    			list[next].add(node);
    		}
    		
    		dfs();
    		
    		for(int i=2; i<=N; i++)
    			System.out.println(parent[i]);
    }
    
    private static void dfs() {
    		Stack<Integer> stack = new Stack<>();
    		visited = new boolean[N+1];
    		
    		stack.push(1);
    		visited[1] = true;
    		
    		while(!stack.isEmpty()) {
    			int node = stack.pop();
    			
    			for(int next: list[node]) {
    				if(!visited[next]) {
    					parent[next] = node;
    					stack.push(next);
    					visited[next] = true;
    				}
    			}
    		}
    }
}